### PR TITLE
api: bmresult POST: document and test that name is required

### PR DESCRIPTION
**Pull request by jgehrcke**
This is a quick follow-up for https://github.com/conbench/conbench/issues/935.

### CODE REVIEW

File: conbench/entities/benchmark_result.py [hunk 1]
```diff
@@ -576,7 +580,31 @@ class _BenchmarkFacadeSchemaCreate(marshmallow.Schema):
...
+                of key/value pairs) indeed corresponds to unique benchmarking
+                behavior. That is, typically, these key/value pairs directly
+                correspond to input parameters to your benchmarking method.
+
+                Example: a conceptual benchmark with name "foo-write-file"
+                might have meaningful case permutations involving tag names
+                such as `compression_method` (values: `gzip`, `lzma`, ...),
+                `file_format` (values: `csv`, `hdf5`, ...), `dataset_name`
+                (values: `foo`, `bar`, ...). For each conceptual benchmark, it
+                is OK to have no or many case permutations.
```
Comment (jgehrcke) [inline conbench/entities/benchmark_result.py:605]: These paragraphs are meant to be provocative! Would love to get your feedback.

Based on https://github.com/conbench/conbench/issues/787 and https://github.com/conbench/conbench/pull/875 (and other discussion threads) I think we're already converging to something.

File: conbench/tests/api/test_benchmarks.py [hunk 2]
```diff
@@ -936,6 +936,16 @@ def test_create_benchmark_context_missing(self, client):
...
 
+    def test_create_benchmark_name_missing(self, client):
+        self.authenticate(client)
+        payload = _fixtures.VALID_PAYLOAD.copy()
+        del payload["tags"]["name"]
+        resp = client.post(self.url, json=payload)
+        # TODO: https://github.com/conbench/conbench/issues/935
+        # This here just quickly checks that there is a failure at all,
+        # that 'name' is required.
+        assert resp.status_code == 500
```
Comment (jgehrcke) [inline conbench/tests/api/test_benchmarks.py]: This quickly confirms existence of https://github.com/conbench/conbench/issues/935, and at least locks in the 'name is required!' condition for now. 

I don't expect this test code to live for long.

File: conbench/entities/benchmark_result.py (hunk 1 above)
Comment (austin3dickey) [reply conbench/entities/benchmark_result.py:605]: This makes sense to me based on what we've talked about!

File: conbench/tests/api/_expected_docs.py [hunk 3]
```diff
@@ -1027,7 +1027,7 @@
...
+                        "description": 'The set of key/value pairs that defines a benchmark case permutation. Top-level keys must be strings. Allowed value types are ... (to be specified).  The special key "name" must be provided with a string value: it indicates the name of the conceptual benchmark that was performed for obtaining the result at hand. All case permutations of a conceptual benchmark by definition have this name in common.  We advise that each unique case (as defined by the complete set of key/value pairs) indeed corresponds to unique benchmarking behavior. That is, typically, these key/value pairs directly correspond to input parameters to your benchmarking method.  Example: a conceptual benchmark with name "foo-write-file" might have meaningful case permutations involving tag names such as `compression_method` (values: `gzip`, `lzma`, ...), `file_format` (values: `csv`, `hdf5`, ...), `dataset_name` (values: `foo`, `bar`, ...). For each conceptual benchmark, it is OK to have no or many case permutations.',
```
Comment (austin3dickey) [inline conbench/tests/api/_expected_docs.py:1030]: Does `dedent_rejoin` not support multiple paragraphs, or is that an OpenAPI spec thing? Would this look cleaner with your paragraphs preserved?

**Review by austin3dickey — APPROVED** [review]
nice!

File: conbench/tests/api/_expected_docs.py (hunk 3 above)
Comment (jgehrcke) [reply conbench/tests/api/_expected_docs.py:1030]: > Would this look cleaner with your paragraphs preserved?

Yes, it would! Will look into that separately. Thanks for sharing the opinion, I thought the same.

File: conbench/entities/benchmark_result.py (hunk 1 above)
Comment (jonkeane) [reply conbench/entities/benchmark_result.py:605]: Thanks for this provocative language + pushing us to be precised.

> it is OK to have no or many case permutations.

Functionally speaking this is true (so long as you have `{name: "foo"}` you're fine[^1]. But practically speaking, there always ends up being at least one parameter and thus at least one case (if you have a parameter that has one value you have a single case, yeah?). So I wonder if we might rephrase this to be "It is ok to have few or many case permutations" and possibly also discuss that there's always at least one case (or else you have an conceptual benchmark with no actual instantiations)?

This hopefully will also nudge folks into thinking a little bit about their case permutations up front. One thing that we've struggled a bit with is as a benchmark evolves over time it usually grows more parameters. That's ok and natural, but currently conbench treats each one as if it were totally separate for history and comparison purposes, so there's a bit of friction on that front. We can't (and shouldn't) solve that here, but something to note that if we can get folks to thing "what are the parameters I might use" that's not a bad thing.


[^1]: Though I suspect that there totally might be code that doesn't handle there being no other tags in that dict and will do funny things if that's all you have.

Comment (jgehrcke) [reply conbench/entities/benchmark_result.py:605]: > there's always at least one case (or else you have an conceptual benchmark with no actual instantiations)?

Yes! This must be true (as in: other thinking needs to change so that this _is_ true): one should be able to submit a benchmark result that has only the `name` set -- it's a special case, but it is a case.

I didn't see this feedback before merging (hmm.. sorry!). Will adjust the text. I think this is what I meant with "no case permutations", but in fact this should be called _one_ case, a special, valid case.

For implementation, we should totally cover that with tests, because it might be an interesting edge case: `name` is set, but no other keys set within `tags`.
